tests/test_descriptor.py

```diff
@@ -77,6 +77,47 @@ def _get_anywidget_state(self):
     mock_comm.send.assert_called()
 
 
+def test_descriptor_sends_hmr_update(mock_comm: MagicMock) -> None:
+    """Test that descriptor sends custom anywidget HMR update to front end."""
+
+    esm = "export function render(view) {}"
+    css = "h1 { }"
+
+    class Foo:
+        _repr_mimebundle_ = MimeBundleDescriptor(autodetect_observer=False)
+        value: int = 0
+
+        def _get_anywidget_state(self):
+            return {"value": self.value}
+
+    foo = Foo()
+
+    foo._repr_mimebundle_._send_hmr_update(esm=esm, css=css)
+    mock_comm.send.assert_called_with(
+        data={
+            "method": "update",
+            "state": {"_esm": esm, "_css": css},
+            "buffer_paths": [],
+        },
+        buffers=[],
+    )
+
+    foo._repr_mimebundle_._send_hmr_update(esm=esm)
+    mock_comm.send.assert_called_with(
+        data={"method": "update", "state": {"_esm": esm}, "buffer_paths": []},
+        buffers=[],
+    )
+
+    foo._repr_mimebundle_._send_hmr_update(css=css)
+    mock_comm.send.assert_called_with(
+        data={"method": "update", "state": {"_css": css}, "buffer_paths": []},
+        buffers=[],
+    )
+
+    foo._repr_mimebundle_._send_hmr_update()
+    mock_comm.send.assert_not_called
```

Comment:
@tlambert03 is this the correct way to assert that the send method doesn't get called? Not deeply familiar with `MagicMock`.